apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -56,27 +166,70 @@ def render(element_html: str, data: pl.QuestionData) -> str:
     file_names = sorted(get_file_names_as_array(raw_file_names))
     file_names_json = json.dumps(file_names, allow_nan=False)
 
-    answer_name = get_answer_name(raw_file_names)
+    raw_opt_file_names = pl.get_string_attrib(element, "optional-file-names", "")
+    opt_file_names = sorted(get_file_names_as_array(raw_opt_file_names))
+    opt_file_names_json = json.dumps(opt_file_names, allow_nan=False)
+
+    raw_file_patterns = pl.get_string_attrib(element, "file-patterns", "")
+    file_patterns = sorted(get_file_names_as_array(raw_file_patterns))
+
+    raw_opt_file_patterns = pl.get_string_attrib(element, "optional-file-patterns", "")
+    opt_file_patterns = sorted(get_file_names_as_array(raw_opt_file_patterns))
+
+    # Convert patterns into regular expressions
+    file_regex = [glob_to_regex(f) for f in file_patterns]
+    opt_file_regex = [glob_to_regex(f) for f in opt_file_patterns]
+
+    # Need to send both converted regex and original pattern to client for matching and printing
+    file_regex_json = json.dumps(list(zip(file_regex, file_patterns)), allow_nan=False)
+    opt_file_regex_json = json.dumps(
+        list(zip(opt_file_regex, opt_file_patterns)), allow_nan=False
+    )
+
+    answer_name = get_answer_name(
+        raw_file_names,
+        raw_opt_file_names,
+        raw_file_patterns,
+        raw_opt_file_patterns,
+    )
+    parse_error = data["format_errors"].get(answer_name, [])
 
     # Only send the file names to the client. We don't include the contents
     # to avoid bloating the HTML. The client will fetch any submitted files
-    # asynchronously once the page loads.
-    #
-    # We filter out any files that weren't specified in the file names for this element.
+    # asynchronously once the page loads
     submitted_files = data["submitted_answers"].get("_files", [])
-    submitted_file_names = list(
-        {x.get("name") for x in submitted_files} & set(file_names)
+    submitted_file_names = [x.get("name") for x in submitted_files]
+
+    # We find any files that match an accepted file name/pattern
+    required_files = set(submitted_file_names) & set(file_names)
+    remaining_files = [x for x in submitted_file_names if x not in file_names]
+    wildcard_files = set(
+        match_regex_with_files(file_regex, remaining_files, limit_1=True)[0]
+    )
+    remaining_files = [x for x in remaining_files if x not in file_names]
```

Comment:
Should this be?
```suggestion
    remaining_files = [x for x in remaining_files if x not in wildcard_files]
```

Reply:
Technically, these two edits don't matter because the sets get merged a few lines later and duplicates wouldn't affect the final result (other than for the 1:1 matching in line 206/207). However, I agree that your edits make the code cleaner and easier to understand.

Reply:
I noticed that, but fixing this future-proofs the code, since it avoids a problem where these variables could be reused in future changes but have invalid values.